README
NAME
    Finance::Bitcoin - manage a bitcoin instance

DESCRIPTION
    Bitcoin is a peer-to-peer network based digital currency.

    This module provides high and low level APIs for managing a running
    bitcoin instance over JSON-RPC.

BUGS
    Please report any bugs to <http://rt.cpan.org/>.

SEE ALSO
    Finance::Bitcoin::API, Finance::Bitcoin::Wallet,
    Finance::Bitcoin::Address.

    <http://www.bitcoin.org/>.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT
    Copyright 2010, 2011, 2013 Toby Inkster

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Finance-Bitcoin
===============

Created:      2010-07-26
Home page:    <https://metacpan.org/release/Finance-Bitcoin>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Finance-Bitcoin>
Maintainer:   Toby Inkster <mailto:tobyink@cpan.org>

0.901  2013-06-16

 - (Bugfix Update) JSON::RPC::Client -> JSON::RPC::Legacy::Client.
   ++$CAVAC

0.900  2013-03-18

 - (Update) Swap from Any::Moose to Moo.
 - General tidy up.

0.004  2011-12-05

 - (Addition) use Object::AUTHORITY
 - Remove stub subs in BEGIN section. (They were only there for the benefit
   of Test::Pod::Coverage, but were causing problems.)
 - Swap from Class::Accessor to Any::Moose.

0.003  2011-10-11

 - (Addition) Include examples in distribution.
 - (Packaging) Module::Package::RDF
 - use 5.010
 - use common::sense

0.002  2010-11-21

 - (Addition) Add $minconf parameter to
   Finance::Bitcoint::Account::received (patch from Pavel Karoukin).

0.001  2010-07-26




lib/Finance/Bitcoin.pm
package Finance::Bitcoin;

use 5.010;
use common::sense;
use Finance::Bitcoin::API;
use Finance::Bitcoin::Wallet;
use Finance::Bitcoin::Address;
use Object::AUTHORITY;

BEGIN {
	$Finance::Bitcoin::AUTHORITY = 'cpan:TOBYINK';
	$Finance::Bitcoin::VERSION   = '0.004';
}

1;

__END__

=head1 NAME

Finance::Bitcoin - manage a bitcoin instance

=head1 DESCRIPTION

Bitcoin is a peer-to-peer network based digital currency.

This module provides high and low level APIs for managing a running
bitcoin instance over JSON-RPC.

=head1 BUGS

Please report any bugs to L<http://rt.cpan.org/>.

=head1 SEE ALSO

L<Finance::Bitcoin::API>,
L<Finance::Bitcoin::Wallet>,
L<Finance::Bitcoin::Address>.

L<http://www.bitcoin.org/>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT

Copyright 2010-2011 Toby Inkster

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



lib/Finance/Bitcoin/API.pm
package Finance::Bitcoin::API;

use 5.010;
use common::sense;
use JSON::RPC::Client;
use Any::Moose;
use Object::AUTHORITY;
use Scalar::Util qw[blessed];

BEGIN {
	$Finance::Bitcoin::API::AUTHORITY = 'cpan:TOBYINK';
	$Finance::Bitcoin::API::VERSION   = '0.004';
}

has endpoint => (is => 'rw', default => sub { 'http://127.0.0.1:8332/' });
has jsonrpc  => (is => 'ro', default => sub { JSON::RPC::Client->new }, lazy => 1 );
has error    => (is => 'rw');

sub call
{
	my ($self, $method, @params) = @_;

	$self->error(undef);
	
	my $return = $self->jsonrpc->call($self->endpoint, {
		method    => $method,
		params    => \@params,
		});
	
	if (blessed($return) and $return->can('is_success') and $return->is_success)
	{
		$self->error(undef);
		return $return->result;
	}
	elsif (blessed($return) and $return->can('error_message'))
	{
		$self->error($return->error_message);
		return;
	}
	else
	{
		$self->error(sprintf('HTTP %s', $self->jsonrpc->status_line));
		return;
	}
}

1;

__END__

=head1 NAME

Finance::Bitcoin::API - wrapper for the Bitcoin JSON-RPC API

=head1 SYNOPSIS

 use Finance::Bitcoin::API;
 
 my $uri     = 'http://user:password@127.0.0.1:8332/';
 my $api     = Finance::Bitcoin::API->new( endpoint => $uri );
 my $balance = $api->call('getbalance');
 print $balance;

=head1 DESCRIPTION

This module provides a low-level API for accessing a running
Bitcoin instance.

=over 4

=item C<< new( %args ) >>

Constructor. %args is a hash of named arguments. You need to provide the
'endpoint' URL as an argument.

=item C<< call( $method, @params ) >>

Call a method. If successful returns the result; otherwise returns undef.

=item C<< endpoint >>

Get/set the endpoint URL.

=item C<< jsonrpc >>

Retrieve a reference to the L<JSON::RPC::Client> object being used. In particular
C<< $api->jsonrpc->ua >> can be useful if you need to alter timeouts or HTTP proxy
settings.

=item C<< error >>

Returns the error message (if any) that resulted from the last C<call>.

=back

=head1 BUGS

Please report any bugs to L<http://rt.cpan.org/>.

=head1 SEE ALSO

L<Finance::Bitcoin>.

L<http://www.bitcoin.org/>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT

Copyright 2010-2011 Toby Inkster

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



lib/Finance/Bitcoin/Address.pm
package Finance::Bitcoin::Address;

use 5.010;
use common::sense;
use Carp;
use Finance::Bitcoin;
use Any::Moose;
use Object::AUTHORITY;
use Scalar::Util qw[blessed];

BEGIN {
	$Finance::Bitcoin::Address::AUTHORITY = 'cpan:TOBYINK';
	$Finance::Bitcoin::Address::VERSION   = '0.004';
}

has address => (is => 'ro');
has api     => (is => 'rw');

around BUILDARGS => sub
{
	my $orig  = shift;
	my $class = shift;

	if (scalar @_ == 2
	and $_[1] =~ /^1/
	and (blessed($_[0]) || $_[0] =~ /^http/))
	{
		my ($api, $address) = @_;

		unless (blessed($api) and $api->isa('Finance::Bitcoin::API'))
		{
			$api = $api ? 
				Finance::Bitcoin::API->new(endpoint=>"$api") : 
				Finance::Bitcoin::API->new;
		}

		return $class->$orig(api => $api, address => $address);
	}
	
	else
	{
		return $class->$orig(@_);
	}
};

sub label
{
	my $self = shift;

	if (@_) # set
	{
		my $label = shift;
		$self->api->call('setlabel', $self->address, $label);
	}

	return $self->api->call('getlabel', $self->address);
}

sub received
{
	my ($self, $minconf) = @_;
	$minconf = 1 unless defined $minconf;
	return $self->api->call('getreceivedbyaddress', $self->address, $minconf);
}

1;

__END__

=head1 NAME

Finance::Bitcoin::Address - a bitcoin address

=head1 SYNOPSIS

 use Finance::Bitcoin;
 
 my $uri     = 'http://user:password@127.0.0.1:8332/';
 my $wallet  = Finance::Bitcoin::Wallet->new($uri);
 
 foreach my $address ($wallet->addresses)
 {
   print $address->address . "\n";
   print $address->label . "\n";
   print $address->received . "\n\n";
 }

=head1 DESCRIPTION

This module is part of the high-level API for accessing a running
Bitcoin instance.

=over 4

=item C<< new($endpoint, $string) >>

Constructor. $endpoint may be the JSON RPC endpoint URL, or may be a
Finance::Bitcoin::API object; $string is an address string.

=item C<< address >>

Returns the address string.

=item C<< label >>

Get/set the address label.

=item C<< received($minconf) >>

Returns the total amount received via this address, with at least $minconf
confirmations. $minconf defaults to 1.

=item C<< api >>

Retrieve a reference to the L<Finance::Bitcoin::API> object being used.

=back

=head1 BUGS

Please report any bugs to L<http://rt.cpan.org/>.

=head1 SEE ALSO

L<Finance::Bitcoin>, L<Finance::Bitcoin::Wallet>.

L<http://www.bitcoin.org/>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT

Copyright 2010-2011 Toby Inkster

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



lib/Finance/Bitcoin/Wallet.pm
package Finance::Bitcoin::Wallet;

use 5.010;
use common::sense;
use Carp;
use Finance::Bitcoin;
use Any::Moose;
use Object::AUTHORITY;
use Scalar::Util qw[blessed];

BEGIN {
	$Finance::Bitcoin::Wallet::AUTHORITY = 'cpan:TOBYINK';
	$Finance::Bitcoin::Wallet::VERSION   = '0.004';
}

has api => (is => 'rw');

around BUILDARGS => sub
{
	my $orig  = shift;
	my $class = shift;

	if (scalar @_ == 1
	and (blessed($_[0]) || $_[0] =~ /^http/))
	{
		my ($api) = @_;

		unless (blessed($api) and $api->isa('Finance::Bitcoin::API'))
		{
			$api = $api ? 
				Finance::Bitcoin::API->new(endpoint=>"$api") : 
				Finance::Bitcoin::API->new;
		}

		return $class->$orig(api => $api);
	}
	
	else
	{
		return $class->$orig(@_);
	}
};

sub balance
{
	my ($self) = @_;
	return $self->api->call('getbalance');
}

sub pay
{
	my ($self, $address, $amount) = @_;

	croak "Must provide an address"
		unless $address;
	croak "Must provide an amount"
		unless $amount;
		
	if (blessed($address))
	{
		$address = $address->address;
	}
	
	return $self->api->call('sendtoaddress', $address, $amount);
}

sub create_address
{
	my ($self, $label) = @_;
	my $address_id = $self->api->call('getnewaddress');
	my $address    = Finance::Bitcoin::Address->new($self->api, $address_id);
	$address->label($label)
		if $label;
	return $address;
}

sub addresses
{
	my ($self) = @_;
	my $list   = $self->api->call('listreceivedbyaddress', 0, JSON::true);
	
	return
		map { Finance::Bitcoin::Address->new($self->api, $_->{address}); }
			grep { $_->{amount} > 0 }
				@$list
		if ref $list eq 'ARRAY';
		
	return;
}

1;

__END__

=head1 NAME

Finance::Bitcoin::Wallet - a bitcoin wallet

=head1 SYNOPSIS

 use Finance::Bitcoin;
 
 my $uri     = 'http://user:password@127.0.0.1:8332/';
 my $wallet  = Finance::Bitcoin::Wallet->new($uri);
 
 print "Have: " . $wallet->balance . "\n";
 $wallet->pay($destination_address, $amount);
 print "Now have: " . $wallet->balance . "\n";
 
 foreach my $address ($wallet->addresses)
 {
   print $address->label . "\n";
 }

=head1 DESCRIPTION

This module is part of the high-level API for accessing a running
Bitcoin instance.

=over 4

=item C<< new($endpoint) >>

Constructor. $endpoint may be the JSON RPC endpoint URL, or may be a
Finance::Bitcoin::API object.

=item C<< balance >>

Returns the current balance of the wallet.

=item C<< pay($dest, $amount) >>

Pays some bitcoins to an account, causing the balance of the wallet to
decrease. $dest may be a Finance::Bitcoin::Address, or an address string.

=item C<< addresses >>

Returns a list of receiving addresses - i.e. addresses that can be used
by other people to send money to this wallet. Each item on the list is
a Finance::Bitcoin::Address object.

This list may be non-exhaustive!

=item C<< create_address($label) >>

Creates a new receiving address - i.e. an address that can be used by
other people to send money to this wallet. $label is an optional
human-friendly name for the address. Returns a Finance::Bitcoin::Address
object.

=item C<< api >>

Retrieve a reference to the L<Finance::Bitcoin::API> object being used. 

=back

=head1 BUGS

Please report any bugs to L<http://rt.cpan.org/>.

=head1 SEE ALSO

L<Finance::Bitcoin>, L<Finance::Bitcoin::Address>.

L<http://www.bitcoin.org/>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT

Copyright 2010-2011 Toby Inkster

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

